File: MAY 24/16 May/Flood fill Algorithm gfg.cpp
class Solution {
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int newColor) {
        // Code here 
        
        int row = image.size();
        int col = image[0].size();
        
        int srcColor = image[sr][sc];
        if(newColor==srcColor)
            return image;
        queue<vector<int>> q;
        q.push({sr, sc});
        image[sr][sc] = newColor;
        int dx[] = {-1,1,0,0};
        int dy[] = {0,0,-1,1};
        vector<int> cell;
        int nx, ny;
        while(!q.empty())
        {
            cell = q.front();
            q.pop();
            
            for(int k=0; k<4; k++)
            {
                nx = cell[0]+dx[k];
                ny = cell[1]+dy[k];
                if(nx<0 || ny<0 || nx>=row || ny>=col || image[nx][ny]!=srcColor)
                    continue;
                image[nx][ny] = newColor;
                q.push({nx, ny});
            }
        }
        return image;
        
    }
};

File: APRIL 2025/12 April/Flood fill Algorithm gfg.java
class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        // Code here
        int original = image[sr][sc];
        int ROW = image.length, COL = image[0].length;
        
        // int[][] visited = new int[ROW][COL];
        
        Queue<int[]> q = new ArrayDeque<>();
        q.offer(new int[] {sr, sc});
        image[sr][sc] = newColor;
        
        while(!q.isEmpty())
        {
            int[] node = q.poll();
            int row = node[0];
            int col = node[1];
            int[][] adjList = {
                {row-1, col}, {row, col+1},{row+1, col},{row, col-1}
            };
            for(int neighbour[]: adjList){
                int r = neighbour[0];
                int c = neighbour[1];
                if(r<0 || r>=ROW || c<0 || c>=COL || image[r][c]!=original || image[r][c] == newColor){
                    continue;
                }
                q.offer(new int[]{r,c});
                image[r][c] = newColor;
            }
        }
        
        return image;
    }
}
